contracts/modules/v2/StakingVe.sol

```diff
@@ -0,0 +1,198 @@
+
+  /* immutables */
+
+  IERC20 public immutable nxm;
+
+  /* constants */
+
+  // 90 * 4 = 360
+  // 91 * 4 = 364
+  // 92 * 4 = 368
+  uint GROUP_SIZE = 91 days;
+
+  constructor (
+    string memory _name,
+    string memory _symbol,
+    IERC20 _token
+  ) ERC721(_name, _symbol) {
+    nxm = _token;
+  }
+
+  // TODO: this should be combined with the processPoolBuckets function
+  function updateGroups() public {
+
+    // SLOAD
+    uint _activeStake = activeStake;
+    uint _rewardPerSecond = rewardPerSecond;
+    uint _accRewardPerToken = accRewardPerToken;
+    uint _poolSharesSupply = poolSharesSupply;
+
+    // SLOAD
+    uint _lastRewardUpdate = lastRewardUpdate;
+    uint _firstActiveGroupId = firstActiveGroupId;
+
+    // the group id for the current timestamp
+    uint currentGroupId = block.timestamp / GROUP_SIZE;
+
+    while (_firstActiveGroupId < currentGroupId) {
+
+      // TODO: check pool buckets expiration here and update _accRewardPerToken & _lastRewardUpdate
+
+      // SLOAD
+      StakeGroup memory group = stakeGroups[_firstActiveGroupId];
+
+      // calculate group reward
+      uint groupExpirationTime = (_firstActiveGroupId + 1) * GROUP_SIZE;
+      uint elapsed = groupExpirationTime - _lastRewardUpdate;
+
+      _accRewardPerToken += elapsed * _rewardPerSecond / _activeStake;
+      _lastRewardUpdate = groupExpirationTime;
+      uint expiredStake = _activeStake * group.poolShares / _poolSharesSupply;
+
+      _activeStake -= expiredStake;
+      _poolSharesSupply -= group.poolShares;
+
+      group.accRewardPerTokenAtExpiration = _accRewardPerToken;
+      group.expiredStakeAmount = expiredStake;
+
+      // SSTORE
+      stakeGroups[_firstActiveGroupId] = group;
+
+      // advance to the next group
+      _firstActiveGroupId++;
+    }
+
+    firstActiveGroupId = _firstActiveGroupId;
+    activeStake = _activeStake;
+  }
+
+  function deposit(
+    uint amount,
+    uint groupId,
+    uint _positionId
+  ) external returns (uint positionId) {
+
+    updateGroups();
+
+    // require groupId not to be expired
+    require(groupId >= firstActiveGroupId);
+
+    // transfer nxm from staker
+    nxm.transferFrom(msg.sender, address(this), amount);
+
+    StakeGroup memory group = stakeGroups[groupId];
+
+    // TODO: double-check that the incresae should happen here
+    uint _activeStake = activeStake + amount;
+    uint _poolSharesSupply = poolSharesSupply;
+
+    uint newPoolShares = _poolSharesSupply == 0
+      ? amount
+      : _poolSharesSupply * amount / _activeStake;
+
+    uint newGroupShares;
+
+    if (group.groupSharesSupply == 0) {
+      newGroupShares = amount;
+    } else {
+      // TODO: the math here is wrong
+      uint groupStake = 0;
+      newGroupShares = group.groupSharesSupply * amount / groupStake;
+    }
+
+    group.groupSharesSupply += newGroupShares;
+    group.poolShares += newPoolShares;
+
+    if (_positionId == 0) {
+      positionId = totalSupply++;
```

Comment:
positionId 0 will never be updateable

Reply:
good catch, it was supposed to be `++totalSupply`